### PR TITLE
Bus support for dma test

### PR DESCRIPTION
**Pull request by trenduchintala**
This adds the capability of specifying which bus is being tested for the fpga_dma_test.

Example of usage: 
fpga_dma_test -B <bus number> <mode 0 or 1>
fpga_dma_test -B 0x04 0

### CODE REVIEW

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 1]
```diff
@@ -54,6 +61,132 @@ static int err_cnt;
...
+		if ((optarg) && ('=' == *tmp_optarg)){
+			++tmp_optarg;
+		}
+		
+		switch (getopt_ret) {
+		case 'B':   /* bus */
+			if (NULL == tmp_optarg)
+				break;
+			endptr = NULL;
+			config.target.bus = (int) strtoul(tmp_optarg, &endptr, 0);
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: You should use safestring versions of strtoul and strlen if they are available

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 2]
```diff
@@ -92,7 +92,7 @@ void help(void)
 /*
  * Parse command line arguments
  */
-#define GETOPT_STRING ":hVNB:D:F:S:AIQ"
+#define GETOPT_STRING "B"
```
Comment (r-rojo) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: I think you need a colon after the B
image

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 3]
```diff
@@ -54,6 +61,132 @@ static int err_cnt;
...
+	
+	int getopt_ret;
+	int option_index;
+	char *endptr = NULL;
+
+	while (-1 != (getopt_ret = getopt_long(argc, argv, GETOPT_STRING, 
+						longopts, &option_index))) {
+		const char *tmp_optarg = optarg;
+		
+		if ((optarg) && ('=' == *tmp_optarg)){
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: Please add a comment

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 4]
```diff
@@ -54,6 +61,132 @@ static int err_cnt;
...
+	
+	return 0;
+}
+
+
+     
+
+
+
+int find_fpga(fpga_guid interface_id, fpga_token *fpga)
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: This function appears to have ambiguous return values. Do you want it to return error code or number of matches? Cant be both.

By convention, all functions should return error code or error count. Use an argument by reference to populate the number of matches.

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 5]
```diff
@@ -161,52 +294,95 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
+	uint8_t bus;
+};
+
+fpga_result get_bus_info(fpga_token tok, struct bus_info *finfo){
+	fpga_result res = FPGA_OK;
+	fpga_properties props;
+	res = fpgaGetProperties(tok, &props);
+	ON_ERR_GOTO(res, out, "reading properties from Token");
+
+        fpgaPropertiesGetBus(props, &finfo->bus);
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: indentation. always use tabs

Comment (deepakcu84) [reply tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: return code isn't checked.

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 6]
```diff
@@ -161,52 +294,95 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
+	ON_ERR_GOTO(res, out_destroy, "Reading bus from properties");
+	
+	out_destroy:
+		fpgaDestroyProperties(&props);
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: return code isnt checked

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 7]
```diff
@@ -161,52 +294,95 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
-
-   res = fpgaPropertiesSetObjectType(filter, FPGA_ACCELERATOR);
-   ON_ERR_GOTO(res, out_destroy_prop, "fpgaPropertiesSetObjectType");
-
-   res = fpgaPropertiesSetGUID(filter, guid);
-   ON_ERR_GOTO(res, out_destroy_prop, "fpgaPropertiesSetGUID");
-
-   res = fpgaEnumerate(&filter, 1, &afc_token, 1, &num_matches);
-   ON_ERR_GOTO(res, out_destroy_prop, "fpgaEnumerate");
+   res = find_fpga(guid, &afc_token);
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: see my prev. comment. res should hold the return code. number of matches should be received in arg.

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 8]
```diff
@@ -161,52 +294,95 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
+   
+   if (res == 0) {
+	fprintf(stderr, "No suitable slots found.\n");
+	goto out_exit;
+	}
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: indent

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 9]
```diff
@@ -161,52 +294,95 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
+   if (res > 1) {
+	fprintf(stderr, "Found more than one suitable slot. ");
+	res = get_bus_info(afc_token, &info);
+        ON_ERR_GOTO(res, out, "getting bus num");
+	print_bus_info(&info);
+	}
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: indent

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 10]
```diff
@@ -161,52 +294,95 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
-   if (num_matches < 1) {
-      printf("Error: Number of matches < 1");
-      ON_ERR_GOTO(FPGA_INVALID_PARAM, out_destroy_prop, "num_matches<1");
-   }
+	
+  if (num_matches < 1) { 
+     printf("Error: Number of matches < 1");
+     ON_ERR_GOTO(FPGA_INVALID_PARAM, out, "num_matches<1");
+  }
+  else if (num_matches > 1) {
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: you dont need the "else if" here.

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 11]
```diff
@@ -54,6 +61,132 @@ static int err_cnt;
...
+"\n"
+                   );
+}
+
+
+/*
+ * Parse command line arguments
+ */
+#define GETOPT_STRING "B"
+int parse_args(int argc, char *argv[])
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: Use fpga_result as the return code. Return FPGA_OK on success, FPGA_EXCEPTION on failure

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 12]
```diff
@@ -300,12 +476,10 @@ int main(int argc, char *argv[])
...
 out_destroy_tok:
    res = fpgaDestroyToken(&afc_token);
-   ON_ERR_GOTO(res, out_destroy_prop, "fpgaDestroyToken");
-
-out_destroy_prop:
-   res = fpgaDestroyProperties(&filter);
-   ON_ERR_GOTO(res, out, "fpgaDestroyProperties");
+   ON_ERR_GOTO(res, out, "fpgaDestroyToken");
 
+out_exit:
```
Comment (deepakcu84) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: see my prev. comment. You don't need to use retval. Use a consistent error type (fpga_result) throughout the application

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c (hunk 1 above)
Comment (trenduchintala) [reply tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: I couldn't find the safestring version of strtoul but I did find the safestring version of strlen which is strnlen. Do you have any other suggestions for strtoul?

Comment (deepakcu84) [reply tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: Ok. No problem. Use strtoul in that case.

**Comment by deepakcu84** [comment]
Looks reasonable. Please remove dead code (code that is commented out).

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 13]
```diff
@@ -161,52 +298,91 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
       return 1;
    }
+
+   res = parse_args(argc, argv);
+   if (res == FPGA_EXCEPTION){
+	goto out_exit;
+   }
    use_ase = atoi(argv[1]);
-   if (use_ase) {
+   if (use_ase) {
```
Comment (asgibbons) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: Space at eol.

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 14]
```diff
@@ -161,52 +298,91 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
+   res = parse_args(argc, argv);
+   if (res == FPGA_EXCEPTION){
+	goto out_exit;
+   }
    use_ase = atoi(argv[1]);
-   if (use_ase) {
+   if (use_ase) { 
       printf("Running test in ASE mode\n");
-   } else {
+   } else if (use_ase == 0) {
```
Comment (asgibbons) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: Seems unnecessarily redundant.

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 15]
```diff
@@ -161,52 +298,91 @@ fpga_result ddr_sweep(fpga_dma_handle dma_h)
...
       printf("Running test in HW mode\n");
    }
-
-   // enumerate the afc
+   
    if (uuid_parse(HELLO_AFU_ID, guid) < 0) {
-      return 1;
+      return FPGA_OK;
```
Comment (asgibbons) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: Should we return OK here?  I thought this was an error.

File: tools/extra/pac/fpgabist/dma/fpga_dma_test.c [hunk 16]
```diff
@@ -345,43 +330,37 @@ int main(int argc, char *argv[])
...
 	goto out_exit;
    }
    use_ase = atoi(argv[1]);
-   if (use_ase) { 
-      printf("Running test in ASE mode\n");
-   } else if (use_ase == 0) {
+   if (use_ase == 0) {
+      printf("Running test in HW mode\n");
+   } else {
       printf("Running test in HW mode\n");
```
Comment (asgibbons) [inline tools/extra/pac/fpgabist/dma/fpga_dma_test.c]: This should print "Running test in ASE mode"

**Pull request merged**